DBZ-3227 Describe how to use predicates to selectively apply SMTs

roldanbob opened the pull request:
[DBZ-3227 ](https://issues.redhat.com/browse/DBZ-3227)

This change adds a new topic file with information about how to use predicates to conditionally apply SMTs.  
Tested in a local Antora build on 02 AUG.

The content is required downstream to complete the work to promote the Outbox SMT to GA ([DBZ-3569](https://issues.redhat.com/browse/DBZ-3569)). 

Please backport this PR to 1.5.

gunnarmorling commented on documentation/modules/ROOT/nav.adoc:
```diff
@@ -12,6 +12,7 @@
 ** xref:configuration/outbox-event-router.adoc[Outbox Event Router]
 ** xref:configuration/filtering.adoc[Message Filtering]
 ** xref:configuration/content-based-routing.adoc[Content-Based Routing]
+** xref:configuration/applying-transformations-selectively.adoc[Using SMT predicates to selectively apply transformations]
```
Let's use the same casing style as for the other entries.

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/applying-transformations-selectively.adoc:
```diff
@@ -0,0 +1,143 @@
+// Title: Applying transformations selectively with SMT predicates
+= Applying transformations selectively
+
+You can define a predicate for a single message transformation to specify a condition that event records must match before Kafka Connect applies the transformation.
```
The predicate decides whether the SMT is applied _at all_, so I think this needs rewording.

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/applying-transformations-selectively.adoc:
```diff
@@ -0,0 +1,143 @@
+// Title: About SMT predicates
+== SMT predicates
+
+{prodname} provides several single message transformations (SMTs) that you can use to modify event records before Kafka Connect saves the records to Kafka topics.
```
SMTs can be applied on the source side (i.e. what you describe here) and on the sink side (i.e. when reading events from a Kafka topic, before passing them to the sink connector). Not sure how to describe this here concisely.

gunnarmorling replied:
@roldanbob, so "Kafka Connect saves the records to Kafka topic" still doesn't quite reflect the ability ot use SMTs with sink connectors.

roldanbob replied:
@gunnarmorling In discussing predicates here I want to keep the focus on how they're used with source connectors such as Debezium. But I did add a paragraph further on that describes a bit more about their use with sink connectors:

> The situation is similar for predicates that you define for a sink connector SMT.
> The connector reads messages from a Kafka topic and Kafka Connect evaluates the messages against the predicate condition.
> If a message matches the condition, Kafka Connect applies the transformation and then passes the messages to the sink consumer.

Does that address your concern?

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/applying-transformations-selectively.adoc:
```diff
@@ -0,0 +1,143 @@
+
+However, there can be instances in which you want to apply a transformation selectively, so that it modifies only that subset of change event messages that share a common characteristic.
+For example, you might want to run the transformation only on event messages from a specific table or that include a specific header key.
+In environments that run Kafka 2.6 or greater, you can append a predicate statement to a transformation to instruct Kafka Connect to apply the SMT only to event records that match a specified condition.
```
```suggestion
In environments that run Apache Kafka 2.6 or greater, you can append a predicate statement to a transformation to instruct Kafka Connect to apply the SMT only to event records that match a specified condition.
```

Let's fully qualify it upon the first usage on a page.

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/applying-transformations-selectively.adoc:
```diff
@@ -0,0 +1,143 @@
+However, there can be instances in which you want to apply a transformation selectively, so that it modifies only that subset of change event messages that share a common characteristic.
+For example, you might want to run the transformation only on event messages from a specific table or that include a specific header key.
+In environments that run Kafka 2.6 or greater, you can append a predicate statement to a transformation to instruct Kafka Connect to apply the SMT only to event records that match a specified condition.
+When a predicate is configured, Kafka Connect evaluates each incoming message that it processes against the condition specified in the predicate.
```
I'd remove the "incoming", so to care for both the source and sink application of SMTs.

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/applying-transformations-selectively.adoc:
```diff
@@ -0,0 +1,143 @@
+If the predicate condition is true for the event message, then the SMT is applied.
+Otherwise, the message is passed without any modifications.
+
+After you define a predicate, you can reuse it and apply it to other transforms.
```
```suggestion
After you define a predicate, you can reuse it and apply it to multiple transforms.
```

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/applying-transformations-selectively.adoc:
```diff
@@ -0,0 +1,143 @@
+Otherwise, the message is passed without any modifications.
+
+After you define a predicate, you can reuse it and apply it to other transforms.
+Predicates also include a `negate` option that you can use to invert a predicate so that the predicate condition is applied only to records that do not match the condition that is defined in the predicate statement.
```
```suggestion
Predicates also include a `negate` option that you can use to invert a predicate so that the predicate condition is applied only to records that do _not_ match the condition that is defined in the predicate statement.
```

Perhaps highlight the _not_?

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/applying-transformations-selectively.adoc:
```diff
@@ -0,0 +1,143 @@
+HasHeaderKey:: Specifies a key name in the header in the event message that you want Kafka Connect to evaluate.
+The predicate evaluates to true for any records that include a header key that has the specified name.
+
+RecordIsTombstone:: Matches records in which the key value of the record is `null`, representing a deleted row in the database.
```
```suggestion
RecordIsTombstone:: Matches records in which the value value of the record is `null`, representing a deleted row in the database.
```

roldanbob replied:
@gunnarmorling I think that you meant to suggest something different here. Remove _key_ with no replacement maybe, i.e.:

_Matches records that have a `null` value?_

gunnarmorling replied:
LOL, yes, you're right of course, and your suggestion is good.

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/applying-transformations-selectively.adoc:
```diff
@@ -0,0 +1,143 @@
+The predicate evaluates to true for any records that include a header key that has the specified name.
+
+RecordIsTombstone:: Matches records in which the key value of the record is `null`, representing a deleted row in the database.
+Tombstones are fields in a database that indicate that a record is not physically deleted, but has a logical deletion flag that is set to `true` when when the record is marked for deletion.
```
That's not quite right actually. Tombstones are used in Kafka as an indicator that all the records with a given key can be dropped from a compacted topic. Debezium emits such tombstone events in case of record deletions (in addition to the "d" change event with the old row state), so that all events for that table row can be removed from Kafka (if compaction is enabled for that topic). Also sink connectors oftentimes rely on tombstone records to remove the represented record from a downstream data store (which implies that settings for the topic must be so that the sink connector is guaranteed to receive the tombstone before Kafka removes it during compaction).

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/applying-transformations-selectively.adoc:
```diff
@@ -0,0 +1,143 @@
+
+RecordIsTombstone:: Matches records in which the key value of the record is `null`, representing a deleted row in the database.
+Tombstones are fields in a database that indicate that a record is not physically deleted, but has a logical deletion flag that is set to `true` when when the record is marked for deletion.
+The predicate evaluates to `true` for any record that has a `null` key value.
```
```suggestion
The predicate evaluates to `true` for any record that has a `null` value.
```

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/applying-transformations-selectively.adoc:
```diff
@@ -0,0 +1,143 @@
+RecordIsTombstone:: Matches records in which the key value of the record is `null`, representing a deleted row in the database.
+Tombstones are fields in a database that indicate that a record is not physically deleted, but has a logical deletion flag that is set to `true` when when the record is marked for deletion.
+The predicate evaluates to `true` for any record that has a `null` key value.
+Use this predicate in combination with a filter SMT to remove tombstone records.
```
Probably worth a mention of the `tombstone.on.deletes` option in Debezium which can be set to false, so to not emit tombstones in the first place. A bit more efficient than first producing them and then filtering them out right away. Though it could be useful actually to enable compaction in Kafka (as described above) but filtering out tombstones on the sink side, as some connectors cannot handle them correctly.

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/applying-transformations-selectively.adoc:
```diff
@@ -0,0 +1,143 @@
+
+TopicNameMatches:: A regular expression that specifies the name of a topic that you want Kafka Connect to match.
+The predicate is true for connector records in which the topic name matches the specified regular expression.
+Use to apply an SMT according to the name of the topic.
```
```suggestion
Use this predicate to apply an SMT according to the name of the topic.
```

The next sentence seems redundant?

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/applying-transformations-selectively.adoc:
```diff
@@ -0,0 +1,143 @@
+[id="defining-smt-predicates"]
+== Defining SMT predicates
+
+By default, Kafka Connect applies each single message transformation in the {prodname} connector configuration to every change event record that it receives from {prodname}.
+Beginning with Apache Kafka 2.6, you can define an SMT predicate for a transformation in the connector configuration that controls how Kafka Connect applies the transformation.
+The predicate statement defines the conditions under which Kafka Connect applies the transformation to event records emitted by {prodname}.
+Kafka Connect evaluates the predicate statement and then applies the SMT selectively to the subset of records that match the condition that is defined in the predicate.
```
Seems a bit redundant with what has been said above?

roldanbob replied:
Yeah, true. In the modularized downstream doc, it's possible for readers to land on a page out of context, so for this procedure I intentionally built in a few sentences to orient someone. It's a bit redundant in the upstream context, so maybe best if I conditionalize it as [product] only.

gunnarmorling reviewed (approved):
Some comments inline, but I really like it overall. Good stuff, @roldanbob!

roldanbob commented:
@gunnarmorling Let me know if you have any further comments/questions on this one. I think that I've addressed all of the concerns that you raised previously. Thanks.

roldanbob commented:
@Naros I just realized that Gunnar isn't available this week. I'd like to get this merged in. Could you or someone else take a peek at the changes and merge the PR if no other changes are needed?

jpechane commented:
@roldanbob Good job, I am fine with the changes. Please add the mentioned section to all SMT chapters and I think we are then good to go. Other SMTs are usually hardened against getting no capture data messages but it is still worthy to have it there in case only subset of topics should be transfromed.

roldanbob commented:
> add the mentioned section to all SMT chapters and I think we are then good to go.

@jpechane Commit fcfcd2c adds the section to the following remaining SMT chapters:

- Topic Routing
- Event flattening/New Record State Extraction
- MongoDB New Document State Extraction

roldanbob commented:
@jpechane @gunnarmorling Added one further change after I realized that the sections that I added to the rest of the SMT chapters include incorrect links. It looks as though the two event-flattening SMTs don't provide a `topic.regex` option, is that right?

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/applying-transformations-selectively.adoc:
```diff
@@ -0,0 +1,190 @@
+Even after the earlier messages are removed, the presence of the tombstone event enables consumers to read the topic from the beginning and retrieve the current state of any events.
+Log compaction occurs periodically, with the compaction interval controlled by the link:{link-kafka-docs}#topicconfigs_delete.retention.ms[`delete.retention.ms`] setting for the topic.
++
+In some cases you might want to configure {prodname} so that it does not emit tombstone events.
```
I'd slightly reword this:

> While you can configure Debezium so it does not emit tombstone events to begin with (see ...), it can make sense in some cases to filter out tombstones for specific sink connectors, using an SMT and this predicate. That way, they can be used for their original purpose of log compaction in Kafka itself, while being ignored by sink connectors which cannot process them.

(just the gist, I hope it gets clear what I mean)

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/applying-transformations-selectively.adoc:
```diff
@@ -0,0 +1,190 @@
+
+.Example: TopicNameMatch predicate for the outbox event router transformation
+
+The following example shows a {prodname} connector configuration that applies the outbox event router transformation only to messages that Debezium emits to the Kafka `outbox.event.order` topic.
```
Debezium -> prod name variable.

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/applying-transformations-selectively.adoc:
```diff
@@ -0,0 +1,190 @@
+Some sink connectors or downstream Kafka consumers are unable to process records with a null value.
+To avoid errors in sink connectors that cannot handle tombstone records, you can configure {prodname} to strip out tombstone records before it saves event records to Kafka topic.
+
+If you support downstream consumers that cannot process tombstones, rather than using the `RecordIsTombstone` predicate to remove tombstone messages after {prodname} emits them, you can prevent {prodname} from emitting tombstones in the first place.
```
See my related comment above: I think it should be framed in a way that you can configure things so that tombstones _are_ emitted (allowing log compaction to remove all records of the affected key) but still can skip them on the sink side for connectors which don't expect them.

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/event-flattening.adoc:
```diff
@@ -102,46 +102,46 @@ To provide the needed Kafka record format for consumers, configure the event fla
-* A tombstone record that has the same key as the deleted row and a value of `null`. This record is a marker for Apache Kafka. It indicates that 
-link:https://kafka.apache.org/documentation/#compaction[log compaction] can remove all records that have this key. 
+* A tombstone record that has the same key as the deleted row and a value of `null`. This record is a marker for Apache Kafka. It indicates that
+link:https://kafka.apache.org/documentation/#compaction[log compaction] can remove all records that have this key.
```
Use Kafka docs prefix variable.

gunnarmorling commented on documentation/modules/ROOT/pages/configuration/topic-routing.adoc:
```diff
@@ -133,6 +133,20 @@ transforms.Reroute.key.enforce.uniqueness=false
+You can use one of the following methods to configure the connector to apply the SMT selectively:
+
+* {link-prefix}:{link-smt-predicates}#applying-transformation-selectively[Configure an SMT predicate for the transformation].
+* Use the xref:by-logical-table-router-topic-regex[topic.regex] configuration option for the SMT.
```
I don't understand how this applies the SMT selectively?

gunnarmorling replied:
(same here and elsewhere)

roldanbob replied:
@gunnarmorling OK, maybe I'm misunderstanding something. From an earlier exchange, I thought that prior to the introduction of predicates, we could use the topic.regex property to apply SMTs only to specific messages.

gunnarmorling replied:
Ah, sorry, I got it now. I had forgotten about that explicit `topic.regex` option we have in some of the SMTs. They were introduced for that very purpose at a time where predicates didn't exist yet in Kafka Connect. The suggested wording makes sense. Thanks!

gunnarmorling reviewed (commented):
@roldanbob, did another review pass; I think once those are addressed, we're good to go with this one.

> It looks as though the two event-flattening SMTs don't provide a topic.regex option, is that right?

Yes, that's right. That's why the topic name predicate is so useful for making sure they are not applied to "special" topics like the ones for TX metadata or heartbeats.

roldanbob commented on documentation/modules/ROOT/pages/configuration/applying-transformations-selectively.adoc:
```diff
@@ -0,0 +1,190 @@
++
+The tombstone represents a delete marker for the the row.
+When link:{link-kafka-docs}#compaction[log compaction] is enabled for Kafka, the presence of a tombstone causes Kafka to remove all events with the same key when it compacts the topic.
+Even after the earlier messages are removed, the presence of the tombstone event enables consumers to read the topic from the beginning and retrieve the current state of any events.
```
> This sentence isn't quite right. It seems it's talking about compaction in general, where the last record with a given key remains in the topic. I don't think it's needed here.

Do you mean just this last sentence?

roldanbob replied:
Oh, nemmind. I see your next comment.

gunnarmorling replied:
This one still is pending, right, @roldanbob?

roldanbob replied:
Yes, @gunnarmorling. Now that you've cleared the path on the question about `topic.regex`, I'll put my other changes through. Thanks.

roldanbob replied:
@gunnarmorling OK, please take a look now. I think that I've addressed all of the comments. Let me know whether it looks ok. Thanks.

gunnarmorling reviewed (approved):
Great work, @roldanbob! I really like how you describe the tombstone semantics, well done 👍 !

roldanbob commented:
> Great work, @roldanbob! I really like how you describe the tombstone semantics, well done  !
Thanks, @gunnarmorling.
So is this all set to merge?

gunnarmorling commented:
> So is this all set to merge?

Yes, it is. I'll just squash the commits a bit and then apply it for 1.5, 1.6, and 1.7.

gunnarmorling commented:
And applied. Thx again, @roldanbob!
